src/content/reference/rules/components-and-hooks-must-be-pure.md

```diff
@@ -0,0 +1,282 @@
+---
+title: Components and Hooks must be pure
+---
+
+<Intro>
+[Purity](/learn/keeping-components-pure) makes your code easier to understand, debug, and allows React to automatically optimize your components and hooks correctly.
+</Intro>
+
+<InlineToc />
+
+<DeepDive>
+#### Why does render need to be pure? {/*why-does-render-need-to-be-pure*/}
+UI libraries like React take care of when your code runs for you so that your application has a great user experience. React is declarative: you tell React what to render in your component's logic, and React will figure out how best to display it to your user!
+
+When render is kept pure, React can understand how to prioritize which updates are most important for the user to see first. This is made possible because of render purity: since components don't have side effects in render, React can pause rendering components that aren't as important to update, and only come back to them later when it's needed.
+
+Concretely, this means that rendering logic can be run multiple times in a way that allows React to give your user a pleasant user experience. However, if your component has an untracked side effect – like modifying the value of a global variable during render – when React runs your rendering code again, your side effects will be triggered in a way that won't match what you want. This often leads to unexpected bugs that can degrade how your users experience your app.
```

Comment:
This is the deep dive section though, so presumably someone's reading this because they want to know more.

> Concretely, this means that rendering logic can be run multiple times in a way that allows React to give your user a pleasant user experience.

this sets the background for why pure components are needed: they enable concurrency (although I've purposefully avoided using the term)

> This is a pretty confusing for someone who hasn't encountered this before. Maybe add an example showing a side effect leading to a undesirable outcome?

there already is an example in the [Keeping Components Pure learn page](http://localhost:3000/learn/keeping-components-pure#side-effects-unintended-consequences). I think there's quite a lot of overlap between these 2 pages, but what I'm imagining the purpose of this page is more of a scannable reference for people who are debugging or in the middle of building something. I added a note to the top of this reference page to first read Keeping Components Pure before this page, so hopefully that gives people the required background